src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/RefInitializationHoister.cs

```diff
@@ -0,0 +1,240 @@
+                        array.Expression,
+                        assignedLocal,
+                        refKind: RefKind.None,
+                        sideEffects,
+                        hoistedSymbols,
+                        ref needsSacrificialEvaluation,
+                        hoister,
+                        createHoistedAccess,
+                        arg,
+                        isRuntimeAsync,
+                        isFieldAccessOfStruct: false);
+
+                    var indices = ArrayBuilder<BoundExpression>.GetInstance();
+                    foreach (var index in array.Indices)
+                    {
+                        indices.Add(HoistExpression(
+                            index,
+                            assignedLocal,
+                            refKind: RefKind.None,
+                            sideEffects,
+                            hoistedSymbols,
+                            ref needsSacrificialEvaluation,
+                            hoister,
+                            createHoistedAccess,
+                            arg,
+                            isRuntimeAsync,
+                            isFieldAccessOfStruct: false));
+                    }
+
+                    needsSacrificialEvaluation = true; // need to force array index out of bounds exceptions
+                    return array.Update(expression, indices.ToImmutableAndFree(), array.Type);
+                }
+
+            case BoundKind.FieldAccess:
+                {
+                    var field = (BoundFieldAccess)expr;
+                    if (field.FieldSymbol.IsStatic)
+                    {
+                        // the address of a static field, and the value of a readonly static field, is stable
+                        if (refKind != RefKind.None || field.FieldSymbol.IsReadOnly) return expr;
+                        goto default;
+                    }
+                    Debug.Assert(field.ReceiverOpt != null);
+
+                    if (refKind == RefKind.None)
+                    {
+                        goto default;
+                    }
+
+                    var isFieldOfStruct = !field.FieldSymbol.ContainingType.IsReferenceType;
+
+                    var receiver = HoistExpression(
+                        field.ReceiverOpt,
+                        assignedLocal,
+                        refKind: isFieldOfStruct ? refKind : RefKind.None,
+                        sideEffects,
+                        hoistedSymbols,
+                        ref needsSacrificialEvaluation,
+                        hoister,
+                        createHoistedAccess,
+                        arg,
+                        isRuntimeAsync,
+                        isFieldAccessOfStruct: isFieldOfStruct);
+                    if (receiver.Kind != BoundKind.ThisReference && !isFieldOfStruct)
+                    {
+                        needsSacrificialEvaluation = true; // need the null check in field receiver
```

Comment:
Do we have a test for this? (I assume it's talking about an NRE)

Reply:
`KeepLtrSemantics_StructFieldAccessOnStructFieldAccessOnClassField` and `KeepLtrSemantics_ClassPropertyAssignmentOnClassProperty`

Reply:
> If I understand correctly, the point is that the NRE should be thrown before evaluating the RHS in arr[index].field = await M();.

That's correct. I can add a comment.